Instance segmentation examples

qubvel opened the pull request:
# What does this PR do?

Add examples that demonstrate how to train MaskFormer and Mask2Former for instance segmentation with Trainer or Accelerate.

Fixes #30335 
(processing void segmentation maps, I can make a separate PR, but for now its included)

At the moment kind of blocked by:
 - #30799 (not that critical, could be merged as is, as a workaround, I pass both values `reduce_labels` and `do_reduce_labels`)
 - #30943 (critical for instance segmentation, because masks are heavy and quickly occupy all RAM)

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

@amyeroberts @NielsRogge

qubvel committed:
Initial setup

qubvel committed:
Metrics

qubvel committed:
Overfit on two batches

qubvel committed:
Train 40 epochs

qubvel committed:
Memory leak debugging

qubvel committed:
Trainer fine-tuning

qubvel committed:
Draft

qubvel committed:
Fixup

qubvel committed:
Trained end-to-end

qubvel committed:
Add requirements

qubvel committed:
Rewrite evaluator

qubvel committed:
nits

qubvel committed:
Add readme

qubvel committed:
Add instance-segmentation to the table

qubvel committed:
Support void masks

qubvel committed:
Remove sh

qubvel committed:
Update docs

qubvel committed:
Add pytorch test

qubvel committed:
Add accelerate test

qubvel commented on examples/pytorch/instance-segmentation/README.md:
```diff
@@ -0,0 +1,239 @@
+
+## Reload and perform inference
+
+This means that after training, you can easily load your trained model and perform inference as follows::
```
```suggestion
This means that after training, you can easily load your trained model and perform inference as follows:
```

qubvel commented on examples/pytorch/instance-segmentation/run_instance_segmentation.py:
```diff
@@ -0,0 +1,489 @@
+
+    # # Sending telemetry. Tracking the example usage helps us better allocate resources to maintain them. The
+    # # information sent is the one passed as arguments along with your Python/PyTorch versions.
+    # send_example_telemetry("run_instance_segmentation", args)
```
```suggestion
    send_example_telemetry("run_instance_segmentation", args)
```

qubvel commented on examples/pytorch/instance-segmentation/run_instance_segmentation_no_trainer.py:
```diff
@@ -0,0 +1,757 @@
+
+    # Sending telemetry. Tracking the example usage helps us better allocate resources to maintain them. The
+    # information sent is the one passed as arguments along with your Python/PyTorch versions.
+    # send_example_telemetry("run_instance_segmentation_no_trainer", args)
```
```suggestion
    send_example_telemetry("run_instance_segmentation_no_trainer", args)
```

qubvel commented on examples/pytorch/instance-segmentation/run_instance_segmentation_no_trainer.py:
```diff
@@ -0,0 +1,757 @@
+        do_resize=True,
+        size={"height": args.image_height, "width": args.image_width},
+        do_reduce_labels=args.do_reduce_labels,
+        reduce_labels=args.do_reduce_labels,  # TODO: remove in the future
```
```suggestion
        reduce_labels=args.do_reduce_labels,  # TODO: remove when mask2former support `do_reduce_labels`
```

qubvel commented on examples/pytorch/instance-segmentation/run_instance_segmentation.py:
```diff
@@ -0,0 +1,489 @@
+        do_resize=True,
+        size={"height": args.image_height, "width": args.image_width},
+        do_reduce_labels=args.do_reduce_labels,
+        reduce_labels=args.do_reduce_labels,  # TODO: remove in the future
```
```suggestion
        reduce_labels=args.do_reduce_labels,  # TODO: remove when mask2former support `do_reduce_labels`
```

qubvel committed:
Update examples/pytorch/instance-segmentation/README.md

qubvel committed:
Update examples/pytorch/instance-segmentation/run_instance_segmentation.py

qubvel committed:
Update examples/pytorch/instance-segmentation/run_instance_segmentation_no_trainer.py

qubvel committed:
Update examples/pytorch/instance-segmentation/run_instance_segmentation_no_trainer.py

qubvel committed:
Update examples/pytorch/instance-segmentation/run_instance_segmentation.py

qubvel committed:
Fix consistency oneformer

qubvel committed:
Merge branch 'instance-segmentation-examples' of github.com:qubvel/transformers into instance-segmentation-examples

qubvel committed:
Fix imports

qubvel committed:
Fix imports sort

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_31084). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

qubvel commented on examples/pytorch/instance-segmentation/run_instance_segmentation.py:
```diff
@@ -0,0 +1,489 @@
+        """Collect predictions in a form of list of dictionaries with keys "masks", "labels", "scores"."""
+
+        model_output = ModelOutput(class_queries_logits=prediction_batch[0], masks_queries_logits=prediction_batch[1])
+        post_processed_output = self.image_processor.post_process_instance_segmentation(
```
Hi @SangbumChoi , thanks a lot for taking time to look into this  🤗 
Not sure I got your question, could you provide more details?

SangbumChoi replied:
when the image size is very large such as 2000x3000, I faced this kind of process (such as resizing into original target size) took almost 10s for single image.

I think I need to test this script with large file :)

qubvel replied:
Ok, got it. I observed that evaluation is relatively slow. However, we have to apply postprocessing in order to get metrics. In case it is too slow we can probably set a rare evaluation strategy (specify a large number of steps between evaluations) or completely drop evaluation during the training and rely on loss values only.
Another thing is to test different approaches to speed up this step, e.g. use lower resolution predicted masks, compute on GPU, or find more optimized operations for postprocessing.
Any help and feedback is appreciated 🤗

SangbumChoi replied:
Yeah, that is also what I thought. I just wanted to make sure if there is any methods in find more optimized operations for postprocessing. Thanks for sharing :)

NielsRogge reviewed (approved):
This is great, thanks for adding this!

Feel free to also add a link to the example scripts in the "resources" section of the respective models

qubvel commented on examples/pytorch/instance-segmentation/README.md:
```diff
@@ -0,0 +1,239 @@
+
+Here is a short script demonstrating how to create your own dataset for instance segmentation and push it to the hub:
+
+> Note: Annotations should be represented as 3-channel images (similar to [scene_parsing_150](https://huggingface.co/datasets/zhoubolei/scene_parse_150#instance_segmentation-1) dataset), first channel is semantic-segmentation map with values corresponding to `label2id`, second is instance-segmentation map, where each instance has its unique value, third channel should be empty (filled with zeros) ...
```
That's a fair point, I feel that your suggested way is more common.
But then they have to provide a class-id/name for each binary mask.

I think the best way would be to support various formats which provide different annotation platforms. On the other hand, usually platforms allow to export in different formats. Probably we can select several most common (like coco) and support them across segmentation, detection, instance segmentation.

qubvel committed:
Apply suggestions from code review

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

SangbumChoi commented:
@qubvel Hi pavel, One more question. How about adding DETR in this script since this is also returns instance output?
https://github.com/huggingface/transformers/blob/573565e35a5cc68f6cfb6337f5a93753ab16c65b/src/transformers/models/detr/image_processing_detr.py#L1774

SangbumChoi commented on examples/pytorch/instance-segmentation/run_instance_segmentation.py:
```diff
@@ -0,0 +1,489 @@
+    )
+    logger.info(f"Training/evaluation parameters {training_args}")
+
+    # Load last checkpoint form output_dir if it exists (and we are not overwriting it)
```
```suggestion
    # Load last checkpoint from output_dir if it exists (and we are not overwriting it)
```

qubvel committed:
Update examples/pytorch/instance-segmentation/run_instance_segmentation.py

Co-authored-by: Sangbum Daniel Choi <34004152+SangbumChoi@users.noreply.github.com>

qubvel commented:
> @qubvel Hi pavel, One more question. How about adding DETR in this script since this also returns instance output?

Hi @SangbumChoi it would be nice to have DETR also, but it's not straightforward since DETR's image processor has a different input annotation format and DETR ModelOutput has different fields it will require adding some conversion logic and I guess the example will be overcomplicated. I tried to keep it simple, but still not completely happy with the amount of code written and how much one should know to be able to modify it and adapt it for a particular use-case.

qubvel committed:
Add resources to docs

amyeroberts reviewed (approved):
Very nice! Thanks for working on and adding this!

SangbumChoi commented:
> it would be nice to have DETR also, but it's not straightforward since DETR's image processor has a different input annotation format and DETR ModelOutput has different fields. It will require adding some conversion logic and I guess the example will be overcomplicated. I tried to keep it simple, but still not completely happy with the amount of code written and how much one should know to be able to modify it and adapt it for a particular use-case.

Exactly, I also try to make the logic simple as much as possible but unfortunately DETR has different annotation input :(
In the future, We can make seperate script for DETR only and just get link in here.

qubvel committed:
Merge branch 'instance-segmentation-examples' of github.com:qubvel/transformers into instance-segmentation-examples

qubvel committed:
Update examples/pytorch/instance-segmentation/README.md

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

qubvel committed:
Update examples/pytorch/instance-segmentation/README.md

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

qubvel committed:
Remove explicit model_type argument

qubvel committed:
Fix tests

qubvel committed:
Update readme

qubvel committed:
Merge branch 'instance-segmentation-examples' of github.com:qubvel/transformers into instance-segmentation-examples

qubvel commented on examples/pytorch/instance-segmentation/README.md:
```diff
@@ -0,0 +1,239 @@
+
+## Reload and perform inference
+
+This means that after training, you can easily load your trained model and perform inference as follows:
```
Ok, thanks!

qubvel committed:
Note about other models

qubvel commented on examples/pytorch/instance-segmentation/run_instance_segmentation.py:
```diff
@@ -0,0 +1,489 @@
+        "model": MaskFormerForInstanceSegmentation,
+        "image_processor": MaskFormerImageProcessor,
+    },
+}
```
Got it, fixed 👍

qubvel commented:
Ok, I addressed most of the comments, thank you all for the reviews 🤗 

I am going to merge this:
 - At the moment, the problem with memory leaks is not critical because it occurs only with persistent DataLoader workers, which I believe is not a common case.
 - Support for other architectures (DETR, Conditional DETR) and dataset formats can be added as a follow-up.
